### PR TITLE
feat: replace serializer with extension nomenclature

### LINKED ISSUE #59: SerializerClass vs IOClass
## Discussion

Would it be more idiomatic to have the current `SerializerClass` be replaced in name by `IOClass` but retain the same functionality.

### Arguments

| For replacing with `AbstractSnapshotIO` | For keeping `Serializer` nomenclature |
| --- | --- |
| IO handles the file system, the serialization is just a step in the process | Serializer handles writing and reading from file system |
| IO (input, output) is more encompassing allowing for extensions to functionality like preparing assertion diffs | Serializer is a more familiar concept to most people making the naming appropriate for its function

### [Example](https://github.com/tophat/syrupy/compare/7cde8a7...feat-22)

https://github.com/tophat/syrupy/blob/44b0f6ccdc1036c531a76dfac87ed5ca75d6be51/README.md#L71
https://github.com/tophat/syrupy/blob/44b0f6ccdc1036c531a76dfac87ed5ca75d6be51/src/syrupy/__init__.py#L66-L71
https://github.com/tophat/syrupy/blob/44b0f6ccdc1036c531a76dfac87ed5ca75d6be51/src/syrupy/assertion.py#L71-L75
https://github.com/tophat/syrupy/blob/44b0f6ccdc1036c531a76dfac87ed5ca75d6be51/src/syrupy/assertion.py#L135-L138

### PR DESCRIPTION
**Pull request by noahnu**
## Description

Finalize v0.1.0 interface

## Related Issues

- Closes #59

## Checklist

- [x] This PR has sufficient test coverage.

## Additional Comments

No additional comments.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -55,16 +55,16 @@ These are the cli options exposed to `pytest` by the plugin.
...
-Syrupy comes with a few built-in serializers for you to choose from. You should also feel free to extend the AbstractSnapshotSerializer if your project has a need not captured by one our built-ins.
+Syrupy comes with a few built-in preset configurations for you to choose from. You should also feel free to extend the AbstractSyrupyExtension if your project has a need not captured by one our built-ins.
 
 - **`AmberSnapshotSerializer`**: This is the default serializer which generates `.ambr` files. Serialization of most data types are supported, however non-sortable types such as frozenset are experimental.
 - **`RawSingleSnapshotSerializer`**: Unlike the `AmberSnapshotSerializer`, which groups all tests within a single test file into a singular snapshot file, the Raw Single serializer creates one `.raw` file per test case.
 - **`PNGSnapshotSerializer`**: An extension of the Raw Single serializer, this should be used to produce `.png` files.
 - **`SVGSnapshotSerializer`**: Another extension of Raw Single. This produces `.svg` files from an svg string.
 
-### Advanced Usage, Plugin Support
+### Advanced Usage, Syrupy Plugin Support
```
Comment (iamogbz) [inline README.md]:
```suggestion
### Advanced Usage, Extending Syrupy
```

File: src/syrupy/extensions/base.py [hunk 2]
```diff
@@ -244,3 +210,45 @@ def __diff_line(
...
+
+    @abstractmethod
+    def _discover_snapshots(self, snapshot_file: str) -> "SnapshotFile":
+        """
+        Given a path to a snapshot file, returns all snapshots in the file.
+        Snapshot name is dependent on serializer implementation.
+        """
+        raise NotImplementedError
+
+    def get_snapshot_name(self, index: int = 0) -> str:
```
Comment (iamogbz) [inline src/syrupy/extensions/base.py]: Should this not be in `SnapshotStorage`

File: README.md [hunk 3]
```diff
@@ -55,32 +55,32 @@ These are the cli options exposed to `pytest` by the plugin.
...
 
 ```python
 import pytest
 
 @pytest.fixture
 def snapshot_custom(snapshot):
     return snapshot.with_class(
-        serializer_class=CustomSerializerClass,
+        extension_class=CustomExtensionClass,
     )
```
Comment (noahnu) [inline README.md]: @iamogbz  how about:
```
return snapshot.with_extension(cls=CustomExtensionClass)
```

or something of this sort?

Comment (iamogbz) [reply README.md]: I do like that, but wouldn't that be confusing? As `cls` name, while not reserved is usually used to denote the object class

Comment (noahnu) [reply README.md]: was mainly thinking about the `with_extension` rename. Not sure what to call the arg if a named kwarg.

File: src/syrupy/data.py [hunk 4]
```diff
@@ -68,7 +68,7 @@ def __iter__(self) -> Iterator["Snapshot"]:
 
 
 @attr.s(frozen=True)
-class SnapshotEmptyFile(SnapshotFile):
+class SnapshotEmptyCache(SnapshotCache):
```
Comment (iamogbz) [inline src/syrupy/data.py]:
```suggestion
class SnapshotEmptyCache(SnapshotCache):
    """This is a cache that is known to be empty and thus can be removed"""
```

File: src/syrupy/data.py [hunk 5]
```diff
@@ -77,33 +77,33 @@ def has_snapshots(self) -> bool:
 
 
 @attr.s(frozen=True)
-class SnapshotUnknownFile(SnapshotFile):
+class SnapshotUnknownCache(SnapshotCache):
```
Comment (iamogbz) [inline src/syrupy/data.py]:
```suggestion
class SnapshotUnknownCache(SnapshotCache):
    """This is a cache that exists but no snapshot extension currently in use has claimed"""
```

File: README.md [hunk 6]
```diff
@@ -55,32 +55,30 @@ These are the cli options exposed to `pytest` by the plugin.
...
 
-- **`AmberSnapshotSerializer`**: This is the default serializer which generates `.ambr` files. Serialization of most data types are supported, however non-sortable types such as frozenset are experimental.
-- **`RawSingleSnapshotSerializer`**: Unlike the `AmberSnapshotSerializer`, which groups all tests within a single test file into a singular snapshot file, the Raw Single serializer creates one `.raw` file per test case.
-- **`PNGSnapshotSerializer`**: An extension of the Raw Single serializer, this should be used to produce `.png` files.
-- **`SVGSnapshotSerializer`**: Another extension of Raw Single. This produces `.svg` files from an svg string.
+- **`AmberSnapshotExtension`**: This is the default extension which generates `.ambr` files. Serialization of most data types are supported, however non-sortable types such as frozenset are experimental.
+- **`SingleFileSnapshotExtension`**: Unlike the `AmberSnapshotExtension`, which groups all tests within a single test file into a singular snapshot file, and creates one `.raw` file per test case.
+- **`PNGSnapshotExtension`**: An extension of raw single, this should be used to produce `.png` files.
```
Comment (iamogbz) [inline README.md]:
```suggestion
- **`PNGSnapshotExtension`**: An extension of single file, this should be used to produce `.png` files.
```

File: src/syrupy/extensions/base.py [hunk 7]
```diff
@@ -40,179 +40,177 @@
...
+            snapshot_name = self.get_snapshot_name(index=index)
+            snapshot_data = self._read_snapshot_data_from_location(
+                snapshot_location=snapshot_location, snapshot_name=snapshot_name
+            )
+            if snapshot_data is None:
+                raise SnapshotDoesNotExist()
+            return snapshot_data
         finally:
-            self.post_read(index=index)
+            self._pre_read(index=index)
```
Comment (iamogbz) [inline src/syrupy/extensions/base.py]:
```suggestion
            self._post_read(index=index)
```

Comment (noahnu) [reply src/syrupy/extensions/base.py]: uh oh. Fixed.

Comment (noahnu) [reply src/syrupy/extensions/base.py]: We also don't have tests for this.

**Pull request merged**